### PR TITLE
 Add BIP341 sighash, enabled by 0x20. 

### PR DESCRIPTION
**Pull request by EyeOfPython**
With the 0x20 bit set, switch the sighash algorithm to BIP341. This is enabled on all output types.

- [x] Separate threading ScriptExecutionData into own PR
- [x] Add tests

### CODE REVIEW

File: src/bench/verify_script.cpp [hunk 1]
```diff
@@ -31,7 +31,7 @@ static void VerifyNestedIfScript(benchmark::Bench &bench) {
     bench.run([&] {
         auto stack_copy = stack;
         ScriptExecutionMetrics metrics = {};
-        ScriptExecutionData execdata = {};
```
Comment (schancel) [inline src/bench/verify_script.cpp:34]: Should this be here?

File: src/script/script_exec_data.h [hunk 2]
```diff
@@ -5,19 +5,50 @@
...
+     *                              script).
+     * - In Taproot, this entire struct would not be constructed.
+     */
+    uint256 m_executed_script_hash;
+
+    /**
+     * ScriptExecutionData with the SHA-256 of executed_script and as
+     * codeseparator position.
+     */
+    ScriptExecutionData(const CScript &executed_script,
```
Comment (schancel) [inline src/script/script_exec_data.h:43]: This should have been in the other PR?

File: src/bench/verify_script.cpp (hunk 1 above)
Comment (schancel) [reply src/bench/verify_script.cpp:34]: These changes are also here? https://github.com/cashweb/cashd/pull/112/files

File: src/script/interpreter.cpp [hunk 3]
```diff
@@ -1776,6 +1776,85 @@ bool SignatureHashBIP143(uint256 &sighashOut, const CScript &scriptCode,
...
+bool SignatureHashBIP341(uint256 &hash_out,
+                         const std::optional<ScriptExecutionData> &execdata,
+                         const T &tx_to, uint32_t in_pos,
+                         SigHashType sig_hash_type,
+                         const PrecomputedTransactionData &cache) {
+    uint8_t ext_flag = bool(execdata);
+    uint8_t key_version = 0;
+    assert(in_pos < tx_to.vin.size());
+
+    CHashWriter ss = HASHER_TAPSIGHASH;
```
Comment (schancel) [inline src/script/interpreter.cpp:1791]: This creates a copy, right?

**Review by schancel — APPROVED** [review]

File: src/script/interpreter.cpp (hunk 3 above)
Comment (EyeOfPython) [reply src/script/interpreter.cpp:1791]: It has to, GetSHA256 invalidates the object at the end of the function

File: src/script/interpreter.cpp [hunk 4]
```diff
@@ -1797,23 +1876,31 @@ bool SignatureHash(uint256 &sighashOut,
                                    sigHashType);
```
Comment (schancel) [inline src/script/interpreter.cpp:1876]: The above needs:

```
if (!sigHashType.hasForkId() || !(flags & SCRIPT_ENABLE_SIGHASH_FORKID)) {
```

File: src/script/interpreter.cpp [hunk 5]
```diff
@@ -1797,23 +1876,31 @@ bool SignatureHash(uint256 &sighashOut,
...
     }
 
-    return SignatureHashBIP143(sighashOut, scriptCode, txTo, nIn, sigHashType,
-                               amount, cache);
+    if (sigHashType.hasBIP341()) {
```
Comment (schancel) [inline src/script/interpreter.cpp:1879]: This needs `if (sigHashType.hasBIP341() && sigHashType.hasForkId()) {`

File: src/script/interpreter.cpp (hunk 4 above)
Comment (schancel) [reply src/script/interpreter.cpp:1876]: We also need an error for:

```
This needs if (sigHashType.hasBIP341() && !sigHashType.hasForkId()) {
```

Currently.

Comment (EyeOfPython) [reply src/script/interpreter.cpp:1876]: Too many tests currently rely on `0x20` being legacy

We can refactor that later.

For now, I'd just return false in `SigHashType::isDefined` in those cases.

**Pull request merged**